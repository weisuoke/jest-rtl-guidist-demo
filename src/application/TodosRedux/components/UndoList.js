import React from 'react'

class UndoList extends React.Component {
  constructor() {
    super();
  }

  render() {
    const { list, deleteItem, changeStatus, handleBlur, valueChange } = this.props
    return (
      <div className="undo-list">
        <div className="undo-list-title">
          正在进行
          <div
            className="undo-list-count"
            data-test="count"
          >
            {list.length}
          </div>
        </div>
        <ul className="undo-list-content">
          {
            list.map((item, index) => {
              return (
                <li
                  data-test="list-item"
                  className="undo-list-item"
                  key={`${item}-${index}`}
                  onClick={() => changeStatus(index)}
                >
                  {item.status === 'div' ? item.value : (
                    <input
                      data-test="input"
                      className="undo-list-input"
                      value={item.value}
                      onBlur={() => handleBlur(index)}
                      onChange={(e) => valueChange(index, e.target.value)}
                    />
                  )}
                  <span
                    data-test="delete-item"
                    className="undo-list-delete"
                    onClick={(e) => {
                      e.stopPropagation()
                      deleteItem(index)
                    }}
                  >
                  x
                </span>
                </li>
              )
            })
          }
        </ul>
      </div>
    )
  }
}

export default UndoList