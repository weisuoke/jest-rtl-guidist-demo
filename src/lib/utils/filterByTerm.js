export const filterByTerm = (inputArr, searchTerm) => {
  return inputArr.filter(function (arrayElement) {
    return arrayElement.url.match(searchTerm)
  })
}