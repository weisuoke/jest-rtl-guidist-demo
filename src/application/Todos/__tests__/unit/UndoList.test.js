import React from 'react'
import { shallow } from 'enzyme'
import UndoList from '../../components/UndoList'
import { findTestWrapper } from '../../../../lib/utils/testUtils'

describe("UndoList组件", () => {
  it('组件渲染正常', () => {
    const wrapper = shallow(<UndoList list={[]}/>)
    expect(wrapper).toMatchSnapshot()
  })

  it('未完成列表当数据为空数组时， count 数目为 0，列表无内容', () => {
    const wrapper = shallow(<UndoList list={[]}/>)
    const countElem = findTestWrapper(wrapper, "count")
    const listItems = findTestWrapper(wrapper, "list-item")
    expect(countElem.text()).toEqual("0")
    expect(listItems.length).toEqual(0)
  })

  it('未完成列表当数据有内容时显示数据长度，内容不为空', () => {
    const listData = [
      {
        status: 'div',
        value: '学习Jest'
      },
      {
        status: 'div',
        value: '学习TDD'
      },
      {
        status: 'div',
        value: '学习单元测试'
      }
    ]
    const wrapper = shallow(<UndoList list={listData}/>)
    const countElem = findTestWrapper(wrapper, "count")
    const listItems = findTestWrapper(wrapper, "list-item")
    expect(countElem.text()).toEqual("3")
    expect(listItems.length).toEqual(3)
  })

  it('未完成列表当数据有内容时，要存在删除按钮', () => {
    const listData = [
      {
        status: 'div',
        value: '学习Jest'
      },
      {
        status: 'div',
        value: '学习TDD'
      },
      {
        status: 'div',
        value: '学习单元测试'
      }
    ]
    const wrapper = shallow(<UndoList list={listData}/>)
    const deleteItems = findTestWrapper(wrapper, "delete-item")
    expect(deleteItems.length).toEqual(3)
  })

  it('未完成列表当数据有内容时，点击某个删除按钮，会调用删除方法', () => {
    const listData = [
      {
        status: 'div',
        value: '学习Jest'
      },
      {
        status: 'div',
        value: '学习TDD'
      },
      {
        status: 'div',
        value: '学习单元测试'
      }
    ]
    const fn = jest.fn()
    const index = 1;
    const wrapper = shallow(<UndoList deleteItem={fn} list={listData}/>)
    const deleteItems = findTestWrapper(wrapper, "delete-item")
    deleteItems.at(index).simulate('click', {
      stopPropagation: () => {}
    });
    expect(fn).toHaveBeenCalledWith(index)
  })

  it('当某一项被点击时，触发执行 changeStatus 函数', () => {
    const listData = [
      {
        status: 'div',
        value: '学习Jest'
      },
      {
        status: 'div',
        value: '学习TDD'
      },
      {
        status: 'div',
        value: '学习单元测试'
      }
    ]
    const fn = jest.fn()
    const index = 1;
    const wrapper = shallow(<UndoList changeStatus={fn} list={listData}/>)
    const listItems = findTestWrapper(wrapper, "list-item")
    listItems.at(index).simulate('click');
    expect(fn).toHaveBeenCalledWith(index)
  })

  it('当某一项状态是Input时，展示输入框', () => {
    const listData = [
      {
        status: 'input',
        value: '学习Jest'
      },
      {
        status: 'div',
        value: '学习TDD'
      },
      {
        status: 'div',
        value: '学习单元测试'
      }
    ]
    const wrapper = shallow(<UndoList list={listData}/>)
    const inputItems = findTestWrapper(wrapper, "input")
    expect(inputItems.length).toBe(1)
  })

  it('当某一项input失去焦点时， 触发执行 handleBlur 方法', () => {
    const listData = [
      {
        status: 'input',
        value: '学习Jest'
      },
      {
        status: 'div',
        value: '学习TDD'
      },
      {
        status: 'div',
        value: '学习单元测试'
      }
    ]
    const fn = jest.fn()
    const wrapper = shallow(<UndoList handleBlur={fn} list={listData}/>)
    const inputElem = findTestWrapper(wrapper, "input")
    inputElem.simulate('blur')
    expect(fn).toHaveBeenCalledWith(0)
  })

  it('当某一输入框变更时，触发执行 handleValueChange 方法', () => {
    const listData = [
      {
        status: 'input',
        value: '学习Jest'
      }
    ]
    const value = '学习TDD'
    const fn = jest.fn()
    const wrapper = shallow(<UndoList valueChange={fn} list={listData}/>)
    const inputElem = findTestWrapper(wrapper, "input")
    inputElem.simulate('change', {
      target: {
        value
      }
    })
    expect(fn).toHaveBeenCalledWith(0, value)
  })
})