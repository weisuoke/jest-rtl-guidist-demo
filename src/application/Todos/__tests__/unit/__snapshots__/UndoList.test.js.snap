// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UndoList组件 组件渲染正常 1`] = `
<div
  className="undo-list"
>
  <div
    className="undo-list-title"
  >
    正在进行
    <div
      className="undo-list-count"
      data-test="count"
    >
      0
    </div>
  </div>
  <ul
    className="undo-list-content"
  />
</div>
`;
